pieforproviders/pieforproviders
# [PIE-1357] get attendances from wonderschool

## 💅🏼 What issue does this fix?
Fixes #1357 

This changes the attendance importer to pull from the remote public URL that Wonderschool built for us.  I refactored a parent Remote CSV Importer class out that the S3 Importer inherits from; might have made sense to separate it further but that felt like overkill to me.

When this is deployed I'll schedule it to run every night at 6pm.  It's safe to run multiple times over the same attendances because it uses the attendance_id from the wonderschool CSV to find the attendance if it already exists and update it.  Hopefully these new "edited" fields won't really matter to us; we're just going to use `checked_in_at` and `checked_out_at` as the source of truth, so as long as they update those fields, we're good.

## 🍂 Before You Submit


## 🛷 Deployment

Add a Heroku scheduler to prod once this is on production.

## 🧵 Steps to set up locally

- Get the ENV VAR for the Wonderschool URL from Kate and add it to your `.env`
- Reset your db/re-seed; your Nebraska kids should get Wonderschool IDs, at least one of which should be in the Wonderschool Attendances CSV

## 🧳 Steps to test
- run the rake task
```
bundle exec rails read_wonderschool_necc_attendances
```
- check that some attendances have been added:
```
bundle exec rails console
> Attendance.count
32
```
- your logging should have loads of entries like this: `[20262] 2021-07-20 19:54:55.981 INFO | Child with Wonderschool ID 1111 not in Pie; skipping` - this is expected; with nearly 1000 records we don't want to throw errors for every non-enrolled-in-Pie child who has an attendance.  At some point it might make more sense to delete those records before processing, but for now, logging what we're skipping makes sense.
- run the rake task again: `bundle exec rails read_wonderschool_necc_attendances`
- check that you still have the same number of attendances
```
bundle exec rails console
> Attendance.count
32
```
- errors might be thrown in the AppSignal development environment if a child in our system has a wonderschool ID that matches the CSV, but doesn't have an approval for the time period of the attendance; you can force that circumstance by re-setting your db to remove all records, and updating one of the NE kids' approval to have expired a long time ago:
```
bundle exec rails console
> Child.joins(:business).where(business: { state: "NE" } ).first.approvals.first.update!(effective_on: Time.current - 20.years, expires_on: Time.current - 19.years)
```
- also update the child to use a wonderschool ID you're sure has attendances in the csv
```
bundle exec rails console
> Child.joins(:business).where(business: { state: "NE" } ).first.update!(wonderschool_id: [ID from CSV])
```
- then run the rake task again, and any attendances for that child should throw an error that will get logged in AppSignal in the development environment

@csprayregen - I can set this up to work in Staging, we may or may not get errors in AppSignal, but it's okay if we don't.  I can force one on Staging after you've done QA just to make sure we're getting them logged.
